pkg/restmapper/lazyrestmapper.go

```diff
@@ -10,40 +10,124 @@ import (
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
-	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
-type RESTMapperProvider func(c *rest.Config) (meta.RESTMapper, error)
-
-// NewPartialRestMapperProvider returns configured 'partial' rest mapper provider intended to be used with controller-runtime manager.
-// Takes GroupFilterPredicate as an argument for filtering out APIGroups during discovery procedure.
-func NewPartialRestMapperProvider(groupFilterPredicate GroupFilterPredicate) RESTMapperProvider {
-	partialRESTMapperProvider := func(c *rest.Config) (meta.RESTMapper, error) {
-		drm, err := apiutil.NewDynamicRESTMapper(c,
-			apiutil.WithLazyDiscovery,
-			apiutil.WithCustomMapper(
-				func() (meta.RESTMapper, error) {
-					dc, err := discovery.NewDiscoveryClientForConfig(c)
-					if err != nil {
-						return nil, err
-					}
-
-					groupResources, err := getFilteredAPIGroupResources(dc, groupFilterPredicate)
-					if err != nil {
-						return nil, err
-					}
-
-					return restmapper.NewDiscoveryRESTMapper(groupResources), nil
-				},
-			),
-		)
-		if err != nil {
-			return nil, err
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
+	mapper      meta.RESTMapper
+	client      *discovery.DiscoveryClient
+	knownGroups map[string]any
```

Comment:
https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set

Might using this make code more readable? Just a thought, I'm ok having a map here.